bootstrap/v2/pkg/kfapp/kustomize/kustomize.go

```diff
@@ -577,84 +572,392 @@ func (kustomize *kustomize) writeConfigFile() error {
-	compKustomizationFile := filepath.Join(comp, kftypes.KustomizationFile)
-	base := path.Join(kustomize.Spec.ManifestsRepo, compPath, "base")
-	baseKustomizationFile := filepath.Join(base, kftypes.KustomizationFile)
-	platformOverlay := path.Join(kustomize.Spec.ManifestsRepo, compPath, "overlays", kustomize.Spec.Platform)
-	platformKustomizationFile := filepath.Join(platformOverlay, kftypes.KustomizationFile)
-	if _, err := os.Stat(platformKustomizationFile); err == nil {
-		bases = append(bases, extractSuffix(kustomize.Spec.ManifestsRepo, platformOverlay))
-	} else if _, err := os.Stat(baseKustomizationFile); err == nil {
-		bases = append(bases, extractSuffix(kustomize.Spec.ManifestsRepo, base))
-	} else if _, err := os.Stat(compKustomizationFile); err == nil {
-		bases = append(bases, extractSuffix(kustomize.Spec.ManifestsRepo, comp))
+// GetKustomization will read a kustomization.yaml and return Kustomization type
+func GetKustomization(kustomizationPath string) *types.Kustomization {
+	kustomizationFile := filepath.Join(kustomizationPath, kftypes.KustomizationFile)
+	data, err := ioutil.ReadFile(kustomizationFile)
+	if err != nil {
+		return nil
 	}
-	kustomization := &types.Kustomization{
-		TypeMeta: types.TypeMeta{
-			Kind:       types.KustomizationKind,
-			APIVersion: types.KustomizationVersion,
-		},
-		Bases: bases,
-		CommonLabels: map[string]string{
-			kftypes.DefaultAppLabel: kustomize.Name,
-		},
-		Namespace: kustomize.Namespace,
+	kustomization := &types.Kustomization{}
+	if err = yaml.Unmarshal(data, kustomization); err != nil {
+		return nil
 	}
-	buf, bufErr := yaml.Marshal(kustomization)
-	if bufErr != nil {
-		return bufErr
+	return kustomization
+}
+
+// MergeKustomization will merge the child into the parent
+// if the child has no bases, then the parent just needs to add the child as base
+// otherwise the parent needs to merge with behaviors
+// Multiple overlays are constrained in what they can merge
+// which exclude NamePrefixes, NameSuffixes, CommonLabels, CommonAnnotations.
+// Any of these will generate an error
+func MergeKustomization(compDir string, targetDir string, kfDef *cltypes.KfDef, params []config.NameValue,
+	parent *types.Kustomization, child *types.Kustomization, kustomizationMaps map[MapType]map[string]bool) error {
+
+	paramMap := make(map[string]string)
+	for _, nv := range params {
+		paramMap[nv.Name] = nv.Value
+	}
+	updateParamFiles := func() error {
+		paramFile := filepath.Join(targetDir, kftypes.KustomizationParamFile)
+		if _, err := os.Stat(paramFile); err == nil {
+			params, paramFileErr := readLines(paramFile)
+			if paramFileErr != nil {
+				return &kfapis.KfError{
+					Code:    int(kfapis.INVALID_ARGUMENT),
+					Message: fmt.Sprintf("could not open %v. Error: %v", paramFile, paramFileErr),
+				}
+			}
+			for i, param := range params {
+				paramName := strings.Split(param, "=")[0]
+				if val, ok := paramMap[paramName]; ok && val != "" {
```

Comment:
if `componentParams` has defined namespace and project, paramMap will have the two key, right?
if so, once params.env also defines  namespace and project, it will always be written by paramMap counterpart in priority, which may be different with kfDef.Namespace and kfDef.Spec.Project